pkg/cmd/pulumi/plugin/plugin_install_test.go

```diff
@@ -278,6 +278,37 @@ func TestGetPluginDownloadForMissingPackage(t *testing.T) {
 	})
 }
 
+func TestRegistryIsNotUsedWhenAFileIsSpecified(t *testing.T) {
+	t.Parallel()
+
+	ctx := context.Background()
+
+	cmd := &pluginInstallCmd{
+		diag: diagtest.LogSink(t),
+		pluginGetLatestVersion: func(ps workspace.PluginSpec, ctx context.Context) (*semver.Version, error) {
+			assert.Fail(t, "GetLatestVersion should not have been called")
+			return nil, nil
+		},
+		registry: testMockRegistry{ /* empty registry will fail when used */ },
+		file:     "./pulumi-resource-some-file.tar.gz", // This is a flag: --file
+		installPluginSpec: func(
+			_ context.Context, _ string,
+			install workspace.PluginSpec, file string,
+			sink diag.Sink, color colors.Colorization, reinstall bool,
+		) error {
+			assert.Equal(t, "./pulumi-resource-some-file.tar.gz", file)
+			assert.Equal(t, workspace.PluginSpec{
+				Name:    "some-file",
+				Kind:    "resource",
+				Version: &semver.Version{Major: 1},
+			}, install)
+			return nil
+		},
+	}
+
+	assert.NoError(t, cmd.Run(ctx, []string{"resource", "some-file", "v1.0.0"}))
```

Comment:
```suggestion
			require.Equal(t, "./pulumi-resource-some-file.tar.gz", file)
			require.Equal(t, workspace.PluginSpec{
				Name:    "some-file",
				Kind:    "resource",
				Version: &semver.Version{Major: 1},
			}, install)
			return nil
		},
	}

	require.NoError(t, cmd.Run(ctx, []string{"resource", "some-file", "v1.0.0"}))
```

We prefer using `require` everywhere since output from `assert.*` is often hard to read, as errors further down the line often don't make sense anymore when earlier commands failed.

Reply:
The logic I go by is use require if the following asserts are guartneed to fail or be meaningless if the current check fails. Else assert is fine. e.g. you nearly _always_ want require.NoError rather than assert.NoError.

Reply:
Me too. Each of the `assert`s I had were independent, and I'd have been happy to see any combination of them together.